### ISSUE TITLE
Testing samples and releases

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#260]
**tl;dr: How to test releases (bundles) and perform integration tests using the set of samples we'll create for CKEditor 5 packages. And much more...**

## Goals

The goal is to enhance our testing suite with:

* integration tests (so far we're mainly writing unit tests),
* mechanisms which would allow testing CKEditor bundles that we'll be officially shipping (there will be many kind of builds and it sometimes happen that something fail in a build, while it work in a dev version).

At the same time we know that we'll have a set of samples (most packages will have some) and we
want to cover these samples by tests too.

## The Idea

All samples of the editor that we'll use on our websites and in the documentation will come from the samples that we'll keep in the repos and which we'll cover by tests.

This means that every important functionality will have its integration tests in form of that sample. Also, these samples will have a form which may allow testing them using a bundled editor.

So the idea is to **use the samples and heir testing mechanisms for testing releases**. This should provide us a sufficient coverage level.

### Sample + Tests

A typical sample JS file will look like this:

```js
ClassicEditor.create( document.querySelector( '#editor' ), {
    features: [ 'paragraph', 'typing', ... ],
    toolbar: [ ... ]
} )
.then( function( editor ) {
    window.editor = editor;
} )
.catch( function( err ) {
    console.error( err.stack );
} );
```

In order to run this as a test we need to import `ClassicEditor`. The `import` statement should not be present in a sample's code in documentation, so it needs to be removed when building docs. To avoid magic, we can mark it with some comment:

```js
import ClassicEditor from '/ckeditor5/editor-classic/classic.js'; // docs:remove

ClassicEditor.create( document.querySelector( '#editor' ), {
    features: [ 'paragraph', 'typing', ... ],
    toolbar: [ ... ]
} )
.then( function( editor ) {
    window.editor = editor;
} )
.catch( function( err ) {
    console.error( err.stack );
} );
```

Then, we need to add code of the tests, but:

* tests need to wait until the editor is ready,
* tests should fail if the editor hasn't been initialised correctly.

Surprisingly, these points are tricky because the code needs to stay simple for the docs purposes and needs to accommodate some additional code for tests.

One of the ideas we had was to do this:

```js
import ClassicEditor from '/ckeditor5/editor-classic/classic.js';

before( function() {
    // docs:start
    return ClassicEditor.create( document.querySelector( '#editor' ), {
        features: [ 'paragraph', 'typing', ... ],
        toolbar: [ ... ]
    } )
    .then( function( editor ) {
        window.editor = editor;
    } )
    .catch( function( err ) {
        console.error( err.stack );
        return err; // docs:remove
    } );
    // docs:end
} );
```

It looks ok, but the ugly part is removing the `return` statement. The code of a sample may not start with it, e.g. if some preparations are needed, so it will be a bit risky regexp.

Another idea is to try to use `bender.defer()`, but I'm not 100% sure that it's going to work:

```js
import ClassicEditor from '/ckeditor5/editor-classic/classic.js';

const done = bender.defer();

// docs:start
ClassicEditor.create( document.querySelector( '#editor' ), {
    features: [ 'paragraph', 'typing', ... ],
    toolbar: [ ... ]
} )
.then( function( editor ) {
    window.editor = editor;
    done(); // docs:remove
} )
.catch( function( err ) {
    console.error( err.stack );
} );
// docs:end
```

In such case, if a promise failed, then the test will timeout, because it will never be unblocked, which should be fine.

### Bundling

There are couple of places in which bundles must work.

The code discussed above will work without any additional processing in normal tests. However, we'd also like to run those tests with bundled editors.

The same requirement come from the docs and from any other production use of the sample.

Last but not least, we need to be able to push samples to JSFiddle. This requires CKEditor bundles to be available somewhere on CDN. They will be used on JSFiddle in the same way as in our docs and on our websites.

#### Docs, Websites and JSFiddle

Currently, we use Rollup for bundling the code. It's configured to expose a global variable under the name that we chose. So after extracting code between `docs:start` and `docs:end` markers we just need to ensure that the bundle is loaded with a `<script>` tag.

#### Tests

This bit is unclear still. Most likely we'll need a separate Bender configuration for running those tests. Bender will inject a `<script>` tag with a needed bundle and the rest of the code should work without any additional processing.

### Types of Bundles

There will be multiple creators (currently we have only `ClassicCreator`) and the current bundler can automatically export only one creator to the global scope.

We could improve the bundler so its exports one file with multiple creators (as an object), but we'd like to avoid that, because of the code size of such bundle. Hence, it's better to generate separate bundle for each creator.

There's even bigger problem with features. Each sample will need a different set of them, so either we'll bundle one huge editor with all features and just enable only a subset of them defined in each sample. Or we can bundle each sample separately. The latter sounds much better, because we'll have super optimised bundles, which will be perfect for usage on our websites. They only need to be published on CDN... and generating them will take ages.

#### Dev and Prod Modes

Documentation builder needs to know paths to the bundles on CDN to be able to generate samples and configuration for JSFiddle. The best thing we can do here is to fully automate that and integrate building the documentation with publishing it and with publishing bundles to CDN. The generated samples will also work for demos on our website

However, we also need to be able to build the documentation quickly, for local purposes. This means that we need a dev mode in which the bundles won't need to be minified, published and perhaps... we don't even need to bundle anything in the dev mode. Since we're able to run tests of the samples without bundling anything, we should be able to include those samples in documentation without bundling too (using AMD).

## Summary

It seems that we should be able to kill multiple birds with one stone here. Use sample tests as integration tests, test releases, produce super optimised builds for our websites and online documentation, have all demos used anywhere tested with each release, etc.

We only need to make sure that all this will work...

**Comment by Reinmar** [comment]
I haven't explained how doc builder could learn what features and creator are needed, but that should be just a matter of some meta comments in samples.

**Comment by pomek** [comment]
If I understood correctly, we have to find solution which allows us to test Samples in 2 environments:
- development mode (scripts will be loaded from local paths),
- production mode  (scripts will be loaded from CDN).

## Whole Sample file

At this moment we have prepared partial solution for above problem. Sample file represents sample for documentation and test suite for itself. Please look at code below, which shows how sample file looks:

```js
/**
 * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */

/* global console:false, bender: false */

import ClassicEditor from '/ckeditor5/editor-classic/classic.js';
import { getData } from '/tests/engine/_utils/model.js';

const done = bender.defer();

// sample:start
ClassicEditor.create( document.querySelector( '#editor' ), {
	features: [ 'enter', 'typing', 'paragraph', 'undo', 'basic-styles/bold', 'basic-styles/italic' ], // sample:remove
	toolbar: [ 'bold', 'italic', 'undo', 'redo' ]
} )
.then( editor => {
	window.editor = editor;
	done(); // sample:remove
}, err => {
	console.error( err.stack );
} );
// sample:end

describe( 'Basic Styles sample', () => {
	it( 'contains default value', () => {
		const output = '<paragraph><selection italic=true /><$text italic=true>This</$text> is an <$text bold=true>editor</$text> instance...</paragraph>';

		expect( getData( editor.document ) ).to.equal( output );
	} );
} );
```

### Sample on documentation page 

The `features` key which comes from configuration object for method `ClassicEditor.create`, cannot be available on documentation page, because it throws an error. `ClassicEditor` tries to find these features, but they are built inside `ClassicEditor` and it's a trouble. 

So, sample script which will execute on documentation page, looks less more like code below:

```js
ClassicEditor.create( document.querySelector( '#editor' ), {
	toolbar: [ 'bold', 'italic', 'undo', 'redo' ]
} )
.then( editor => {
	window.editor = editor;
}, err => {
	console.error( err.stack );
} );
```

That code is located between comments `// sample:start` and `// sample:end`. Lines which contain comment `// sample:remove` are removed. 

### Test sample in development environment

Generally, code above works good in development build. Sample are testable and executable (on documentation page). I'm afraid that it's not enough. Samples for production need whole editor (variable like `ClassicEditor`) inside test suite file. I guess, I came up with solution.

### Test sample in production environment

Let's assume that documentation builder knows available editors:
```js
docsBuilder( {
    // ...
    editors: {
        ClassicEditor: {
            js: 'path/to/classicEditor.js',
            css: 'path/to/classicEditor.css'
        },
        MarkdownEditor: {
            js: 'path/to/markdownEditor.js',
            css: 'path/to/markdownEditor.css'
        }
    }
} );
```

And each sample has to have defined which editor will be used (it can be solved by meta tags in description of sample markdown file, because it's already prepared).

```
editor: ClassicEditor 

---

Description of amazing sample.
```

When `docsBuilder` is working in production mode:
- all `import`s will be removed,
- the `features` key also has to be removed,
- whole editor definition will be inserted before test suite. 

Sample test script in production mode will look like below:

```js
// Definition of editor
var ClassicEditor = ( function() { /* ... */ } )();

const done = bender.defer(); // sample:remove

ClassicEditor.create( document.querySelector( '#editor' ), {
	toolbar: [ 'bold', 'italic', 'undo', 'redo' ]
} )
.then( editor => {
	window.editor = editor;
	done(); // sample:remove
}, err => {
	console.error( err.stack );
} );

describe( 'Basic Styles sample', () => {
	it( 'contains default value', () => {
		const output = '<paragraph><selection italic=true /><$text italic=true>This</$text> is an <$text bold=true>editor</$text> instance...</paragraph>';

		expect( getData( editor.document ) ).to.equal( output );
	} );
} );
```

It allows to test sample in 2 environments and… avoid to thinking about CDN. Sources for documentation (available editors) will be in the same directory where rest of files, so it will work.  JsFiddle will also work because we will have real links (source will be able on our server).

**Comment by Reinmar** [comment]
> The features key which comes from configuration object for method ClassicEditor.create, cannot be available on documentation page, because it throws an error. ClassicEditor tries to find these features, but they are built inside ClassicEditor and it's a trouble.

It shouldn't throw: https://github.com/ckeditor/ckeditor5-dev-bundler-rollup/issues/1.

**Comment by Reinmar** [comment]
> If I understood correctly, we have to find solution which allows us to test Samples in 2 environments:

That's only a part of what we need to figure out.

> development mode (scripts will be loaded from local paths),

Yes, and most importantly – in dev mode code isn't concatenated.

> production mode (scripts will be loaded from CDN).

Nope, it's not about CDN. It's only about running tests on a build version of CKEditor. The code will be loaded locally, but will be bundled first.

**Comment by Reinmar** [comment]
> Generally, code above works good in development build. Sample are testable and executable (on documentation page). I'm afraid that it's not enough. Samples for production need whole editor (variable like ClassicEditor) inside test suite file.

This can be done by simply inserting a script tag with a specific CKEditor bundle into the page. Bender can do that using some `@bender-include:` comment AFAIR (search for them in CKEditor 4 tests).

> Let's assume that documentation builder knows available editors:

It's not going to know that. This must be done automatically. Remember about CKEditor's plugin architecture. Creators are simple modules which can be defined in one of installed packages. The builder will not know about all installed packages.

But it shouldn't be a problem. For a  test to work in the dev mode, the creator must be imported to the file. So when building tests for running them in a production mode, the builder can tak this path and pass it to the bundler. The bundler will use that creator and spit out a single JS file. Later on just import that file in that test using a `@bender-include` directive.

**Comment by pomek** [comment]
I was thinking that I'm as close as never before. But… a lot of other problems came to me. 

I've spent several days for preparing [module which builds editor based on samples](https://github.com/ckeditor/ckeditor5-dev-bundler-rollup/blob/t/7/lib/index.js#L300) and [module which transforms samples for tests with bundled editors](https://github.com/ckeditor/ckeditor5-dev-builder/blob/t/3/lib/builder/index.js#L362).

Unfortunately, there are a few problems:

- **gulp tasks** -  sometimes `build:samples:tests` start before ending `bundle:samples` and requires files aren't available
- **ckeditor5-dev-builder** - looks bad because a huge part of the code has been copied (in other case throws error or doesn't work well)
- **tests** - there is no way to test sample with bundled editors because tests utils don't work. More details below.

My whole sample looks:

```js
/**
 * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */

/* global console:false, bender: false */

import ClassicEditor from '/ckeditor5/editor-classic/classic.js'; // editor:module
import { getData } from '/tests/engine/_utils/model.js';

const done = bender.defer();

// sample:start
ClassicEditor.create( document.querySelector( '#editor' ), { // editor:name
	features: [ 'enter', 'typing', 'paragraph', 'undo', 'basic-styles/bold', 'basic-styles/italic' ], // sample:remove|editor:features
	toolbar: [ 'bold', 'italic', 'undo', 'redo' ]
} )
.then( editor => {
	window.editor = editor;
	done(); // sample:remove
}, err => {
	console.error( err.stack );
} );
// sample:end

describe( 'Basic Styles sample', () => {
	it( 'contains default value', () => {
		const output = '<paragraph><selection italic=true /><$text italic=true>This</$text> is an <$text bold=true>editor</$text> instance...</paragraph>';
		expect( getData( window.editor.document ) ).to.equal( output );
	} );
} );
```

During transformation sample for AMD module with sample and tests, the string `// sample:start` is replacing on an editor definition. The whole a sample/test is built properly. 

When the test is running, method `getData` throws an error: `TypeError: Document needs to be an instance of engine.model.Document.`.

I tried to remove part of the code which throws the error. Unfortunately, next method throws an error. I abandoned this method to solve the problem. The same test running on local environment works properly.

If anybody wants to reproduce this error, I will help. From my point of view, at this moment the samples are untestable with bundled editors.

**Comment by Reinmar** [comment]
> ```
> .then( editor => {
>     window.editor = editor;
>     done(); // sample:remove
> }, err => {
>     console.error( err.stack );
> } );
> ```

Don't use then's err param in this way. If the first callback fails, you'll not catch the error. So it's better to use explicite `catch()`

**Comment by Reinmar** [comment]
I pushed some test samples to https://github.com/ckeditor/ckeditor5-core/tree/t/ckeditor5/260 so we don't need to create them every time manually. I'm not sure whether creating yet another jshintrc makes sense, but since we can automate their updates thanks to `gulp exec`, perhaps it doesn't matter.

**Comment by Reinmar** [comment]
There are couple of things that I noticed already:

1. The sample tests (not bundled) should be generated together with all other tests (so when using `gulp compile [--watch]`). The goal is to use sample tests as part of integration tests. This, however, may require more coding, so we don't have to work on this in this PR.
2. In the bundled scenario, the editor package is injected directly to test JS file. It should be saved on a disk as a separate file and loaded using `bender-include` directive (like [here](https://github.com/ckeditor/ckeditor-dev/blob/846f2a1a8f45db9b2c59538b59cbbfb1080b84ed/tests/plugins/clipboard/inline.js#L3)).
3. When marking lines (`// sample:remove|editor:features`), I think that comma (followed by a space of course!) would be more suitable than `|` to separate directives.
